### ISSUE TITLE
Error in CMakeLists.txt in 3.19.1: Runtime Library for Visual Studio

### ISSUE THREAD

**Issue by Frank-Schmitt** [protocolbuffers/protobuf#9209]
CMakeLists.txt in 3.19.1 includes the following block which is meant to set if we should use the dll or the static version of the C runtime. The lower block only switches to static runtime if the CMake option protobuf_MSVC_STATIC_RUNTIME is set. In the upper block which is used when a more recent cmake is available, the option protobuf_MSVC_STATIC_RUNTIME is ignored. I propose to change this to allow users of modern cmake versions to specify if they want static or dynamic runtime.

Old code:

```
if (protobuf_BUILD_SHARED_LIBS)
  set(protobuf_SHARED_OR_STATIC "SHARED")
else (protobuf_BUILD_SHARED_LIBS)
  set(protobuf_SHARED_OR_STATIC "STATIC")
  # The CMAKE_<LANG>_FLAGS(_<BUILD_TYPE>)? is meant to be user controlled.
  # Prior to CMake 3.15, the MSVC runtime library was pushed into the same flags
  # making programmatic control difficult.  Prefer the functionality in newer
  # CMake versions when available.
  if(CMAKE_VERSION VERSION_GREATER 3.15 OR CMAKE_VERSION VERSION_EQUAL 3.15)
    set(CMAKE_MSVC_RUNTIME_LIBRARY MultiThreaded$<$<CONFIG:Debug>:Debug>)
  else()
    # In case we are building static libraries, link also the runtime library statically
    # so that MSVCR*.DLL is not required at runtime.
    # https://msdn.microsoft.com/en-us/library/2kzt1wy3.aspx
    # This is achieved by replacing msvc option /MD with /MT and /MDd with /MTd
    # http://www.cmake.org/Wiki/CMake_FAQ#How_can_I_build_my_MSVC_application_with_a_static_runtime.3F
    if (MSVC AND protobuf_MSVC_STATIC_RUNTIME)
      foreach(flag_var
          CMAKE_CXX_FLAGS CMAKE_CXX_FLAGS_DEBUG CMAKE_CXX_FLAGS_RELEASE
          CMAKE_CXX_FLAGS_MINSIZEREL CMAKE_CXX_FLAGS_RELWITHDEBINFO)
        if(${flag_var} MATCHES "/MD")
          string(REGEX REPLACE "/MD" "/MT" ${flag_var} "${${flag_var}}")
        endif(${flag_var} MATCHES "/MD")
      endforeach(flag_var)
    endif (MSVC AND protobuf_MSVC_STATIC_RUNTIME)
  endif()
endif (protobuf_BUILD_SHARED_LIBS)
```

Change could be this:

```
  if(CMAKE_VERSION VERSION_GREATER 3.15 OR CMAKE_VERSION VERSION_EQUAL 3.15 AND protobuf_MSVC_STATIC_RUNTIME)
```

**Comment by JoelKKelso** [comment]
Seconded. I just spent a few hours trying to figure out why my protobuf rebuild had switched to /MTd compared to the last rebuild.
(Does that change above need parenthesis around the OR part ?)

**Comment by jammyworks** [comment]
Upvoted.   Haven't found a way to get MultiThreadedDebugDLL other than uncommenting that entire block and passing -D CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreadedDebugDLL .

**Comment by h-vetinari** [comment]
I believe this was fixed by #9437
